server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java

```diff
@@ -0,0 +1,314 @@
+    poolNameToRegexPattern = new HashMap<>();
+    for (Entry<String,String> table : getTableOperations().tableIdMap().entrySet()) {
+      tableIdToTableName.put(table.getValue(), table.getKey());
+      Map<String,String> customProps = conf.getTableConfiguration(table.getValue()).getAllPropertiesWithPrefix(Property.TABLE_ARBITRARY_PROP_PREFIX);
+      if (null != customProps && customProps.size() > 0) {
+        for (Entry<String,String> customProp : customProps.entrySet()) {
+          String tableName = customProp.getKey().substring(HOST_BALANCER_PREFIX.length());
+          String regex = customProp.getValue();
+          poolNameToRegexPattern.put(tableName, Pattern.compile(regex));
+        }
+      }
+    }
+    String oobProperty = conf.getConfiguration().get(HOST_BALANCER_OOB_CHECK);
+    if (null != oobProperty) {
+      oobCheckMillis = AccumuloConfiguration.getTimeInMillis(oobProperty);
+    }
+    String poolRecheckProperty = conf.getConfiguration().get(HOST_BALANCER_POOL_RECHECK_KEY);
+    if (null != poolRecheckProperty) {
+      poolRecheckMillis = AccumuloConfiguration.getTimeInMillis(poolRecheckProperty);
+    }
+    String ipBased = conf.getConfiguration().get(HOST_BALANCER_REGEX_USING_IPS);
+    if (null != ipBased) {
+      isIpBasedRegex = Boolean.parseBoolean(ipBased);
+    }
+  }
+
+  public Map<String,String> getTableIdToTableName() {
+    return tableIdToTableName;
+  }
+
+  public Map<String,Pattern> getPoolNameToRegexPattern() {
+    return poolNameToRegexPattern;
+  }
+
+  public long getOobCheckMillis() {
+    return oobCheckMillis;
+  }
+
+  public long getPoolRecheckMillis() {
+    return poolRecheckMillis;
+  }
+
+  public boolean isIpBasedRegex() {
+    return isIpBasedRegex;
+  }
+
+  @Override
+  public void init(ServerConfiguration conf) {
+    super.init(conf);
+    parseConfiguration(conf);
+  }
+
+  @Override
+  public void getAssignments(SortedMap<TServerInstance,TabletServerStatus> current, Map<KeyExtent,TServerInstance> unassigned,
+      Map<KeyExtent,TServerInstance> assignments) {
+
+    if ((System.currentTimeMillis() - this.lastOOBCheck) > this.oobCheckMillis) {
+      // Check to see if a tablet is assigned outside the bounds of the pool. If so, migrate it.
+      for (Entry<TServerInstance,TabletServerStatus> e : current.entrySet()) {
+        String assignedPool = getPoolNameForHost(e.getKey().host());
+        for (String pool : poolNameToRegexPattern.keySet()) {
+          if (assignedPool.equals(pool) || pool.equals(DEFAULT_POOL)) {
+            continue;
+          }
+          String tid = getTableOperations().tableIdMap().get(pool);
+          try {
+            List<TabletStats> outOfBoundsTablets = getOnlineTabletsForTable(e.getKey(), tid);
+            for (TabletStats ts : outOfBoundsTablets) {
+              LOG.info("Tablet {} is currently outside the bounds of the regex, reassigning", ts.toString());
+              unassigned.put(new KeyExtent(ts.getExtent()), e.getKey());
```

Comment:
Assuming this is safe for now, you are taking an entry from current, adding it to unassigned, but leaving it in current. Then the results get passed to the table-specific balancer below. Not sure if this will cause any problems for the table-specific balancers. It's probably worth removing the assignment from current.